File: src/leetcode/2024/07-21-24/number-of-islands.ts
/*
    Link: https://leetcode.com/problems/number-of-islands/description/

    Time Took: 23 minutes

    Time Complexity: O(M+N). Must visit every cell in the grid (at worst case twice)

    Space Complexity: O(1), only additonal space needed in the queue

    Solution Explained:

    - Do a BFS when land is found, find all adjacent land and when visited mark as water (as to not visit again.)
*/

const LAND: string = "1";
const WATER: string = "0";

class Cell {
    row: number;
    col: number;

    constructor(row: number, col: number) {
        this.row = row;
        this.col = col;
    }
}

function bfs(grid: string[][], startRow: number, startCol: number): void {
    let queue: Cell[] = [ new Cell(startRow, startCol) ];
    grid[startRow][startCol] = WATER; 

    while(queue.length > 0) {
        const cell: Cell = queue.shift()!;

        const neighbours: Cell[] = [
            new Cell(cell.row, cell.col - 1), // left
            new Cell(cell.row, cell.col + 1), // right
            new Cell(cell.row - 1, cell.col), // up
            new Cell(cell.row + 1, cell.col), // down
        ];

        for (const neighbour of neighbours) {
            if (neighbour.col >= 0 
                && neighbour.col < grid[0].length
                && neighbour.row >= 0 
                && neighbour.row < grid.length
                && grid[neighbour.row][neighbour.col] === LAND
            ) {
                grid[neighbour.row][neighbour.col] = WATER; 
                queue.push(neighbour);
            }
        }
    }
}

function numIslands(grid: string[][]): number {
    let numberOfIslands: number = 0;
    
    for(let row: number = 0; row < grid.length; row++) {
        for (let col: number = 0; col < grid[row].length; col++) {
            if (grid[row][col] === LAND) {
                bfs(grid, row, col);
                numberOfIslands++;
            }
        }
    }

    return numberOfIslands;
}

File: src/leetcode/questions/graph/number-of-islands.py
'''
    Link: https://leetcode.com/problems/number-of-islands/description/

    Time Took: 25 mins (I messed up and forgot to only add to queue if land)

    Time Complexity: O(M x N), scales with size of grid

    Space Complexity: O(L), scales with amount of land, not grid, even with HUGE grid, if so land, no memory used

    Date: 02-02-2025

    Problem Type: graph

    Solution Explained: Use BFS, loop through matrix, when you see land, start BFS. BFS will find all adjacent land, to avoid cycles. looks in all 4 directions to find adjacent land.

    Notes: python while queue: will stop when queue is empty, use deque for a queue, has .append and .popleft, tuples are nice for dirs, since you can loop through easily.
'''

from collections import deque
from typing import List

LAND='1'
WATER='0'

dirs = [
    (0, -1),
    (0, 1),
    (1, 0),
    (-1, 0),
]

class Solution:
    grid = []
    rows = 0
    cols = 0

    def bfs(self, startRow, startCol):
        queue = deque([(startRow, startCol)])

        while queue:
            row, col = queue.popleft()

            for rowShift, colShift in dirs:
                newRow, newCol = row + rowShift, col + colShift

                if newCol >= 0 and newCol < self.cols and newRow >= 0 and newRow < self.rows:
                    if self.grid[newRow][newCol] == LAND:
                        self.grid[newRow][newCol] = WATER
                        queue.append((newRow, newCol))


    def numIslands(self, grid: List[List[str]]) -> int:
        self.grid = grid
        self.rows = len(grid)
        self.cols = len(grid[0])
        islands = 0

        for row in range(self.rows):
            for col in range(self.cols):
                if grid[row][col] == LAND:
                    self.bfs(row, col)
                    islands += 1

        return islands
